### DISCUSSION TITLE
WebAssembly (Wasm) audits in Lighthouse

### DISCUSSION THREAD

**Discussion by tomayac** [GoogleChrome/lighthouse#15325] [Ideas]
There are a number of things which apps that include WebAssembly (Wasm) code can get wrong, and that Lighthouse could warn about:

- When Lighthouse detects a targeted-at-all-environments Wasm file is executed in the browser, print a hint that points at the [`-s ENVIRONMENT`](https://github.com/emscripten-core/emscripten/blob/42d21d4b60c1768be6f3365606bb67c31ff224a0/src/settings.js#L621) option as an optimization opportunity. JavaScript glue code size savings for a "hello world" example: ~5% smaller (`-s ENVIRONMENT=web` vs. regular [[diff](https://gist.github.com/tomayac/7df635ec2edf909d6ee2842be2be5c15#file-wasm-environment-web-vs-all-diff)]).

- When Lighthouse detects an unoptimized Wasm file is executed in the browser, print a hint that points at the [optimization options](https://emscripten.org/docs/optimizing/Optimizing-Code.html#how-to-optimize-code). Wasm code size savings for a "hello world" example: ~83% smaller (`-O3` vs. `-O0`). 

- When Lighthouse detects a Wasm file is delivered to production with the `-g` debug option enabled, show a warning message. Wasm code size savings for a "hello world" example: ~95% smaller (regular vs. `-g`). 

Now these savings of course need to be taken with a huge grain of salt and will vary in actually useful code. Find the source code in the accompanying [testing repo](https://github.com/tomayac/wasm-lighthouse-audit). It's napkin math that doesn't fully demonstrate the savings potential yet, but that could help start a discussion.

**Comment by connorjclark** [comment]
Did we ever find a reasonable approach for attaching the build options used (optimization level, env, etc), or is the idea here to infer the build options by analyzing the end result?

**Comment by tomayac** [comment]
The idea would be to find this out as part of the discussion… Some (internal to Google, sorry) commenting happened in [this doc](https://docs.google.com/document/d/1DHjjpM0k0GkZY-hp_A5o8TW9WyNosqPoIbRiUxQjT6k/edit?resourcekey=0-n2iFu0PKwkDbkTYn7mPeiA#heading=h.58bcms7u9jiv).

**Reply by connorjclark** [reply]
The most relevant comment thread: [here](https://docs.google.com/document/d/1DHjjpM0k0GkZY-hp_A5o8TW9WyNosqPoIbRiUxQjT6k/edit?resourcekey=0-n2iFu0PKwkDbkTYn7mPeiA&disco=AAAAyusiUQs)

**Comment by tomayac** [comment]
FYI: @dschuff, @kripken.

**Comment by tomayac** [comment]
For detecting unoptimized Wasm, @connorjclark and @kripken suggest to run [`wasm-opt`](https://github.com/webassembly/binaryen#wasm-opt) over the Wasm file and to compare the results with what was shipped. This was also the approach taken by the [HTTP Almanac](https://almanac.httparchive.org/en/2021/webassembly#how-much-can-we-save-via-wasm-opt), though with a little unclear outcomes.

**Comment by tomayac** [comment]
For detecting Wasm with debug information, you can run `wasm-opt --strip-debug` and compare to the original file size.

**Comment by tomayac** [comment]
For detecting Wasm that has not been optimized to run exclusively on the Web, you can check for the _non_-existence of the following string (it's _there_ if the library was optimized to run  exclusively on the Web): 

```js
throw new Error('not compiled for this environment (did you build to HTML and try to run it not on the web, or set ENVIRONMENT to something - like node - and run it someplace else - like on the web?)')
```

Not sure how robust this would be, though…

**Reply by kripken** [reply]
Another option here might be to look for any Node.js support code, which is not emitted when `-sENVIRONMENT=web,worker` etc. Some examples in the code, that appear even in minified fully with closure:
```js
require("fs")
require("path")
.readFileSync
.readFile
process.argv
process.on
process.exitCode
```

**Reply by connorjclark** [reply]
Looking for code like this is too much of a moving target. It's good for existing stuff to know how to detect, but for things moving forward we need a better indicator. Can the JS output from emscripten include some global `meta` object with any relative build details? What can we include by default that 1) does not have privacy or security implications and 2) is not unreasonably large?

**Reply by kripken** [reply]
Yeah, I get that a moving target may not be worth handling, definitely a fair point.

We have discussed in the past including build info in Emscripten, but the decision has been negative (at least by default), for reasons including the following:

* Code size, since even a small amount of extra code is a negative for many of our users. Adding more code size by default forces them to set a flag and be aware they need to set it; they really dislike us adding such flags over time.
* Awkward incentives. We as toolchain makers would love to know how much our toolchain is used, but making our users ship larger builds for our benefit feels wrong. (In the context here, the user would benefit, so this is not entirely applicable.)
* Lack of precedent in the JS space. (Has this changed?)
* Poor precedent in the native space. (There actually [is a precedent](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-qn), but at least my investigation could not actually find a good historical reason for it, and plenty of online threads of people surprised and annoyed by it.)

**Reply by connorjclark** [reply]
For the very specific "was this built with NODE support", can we just look for `ENVIRONMENT_IS_NODE` variable being present?

More generally ...

> 1. Code size, since even a small amount of extra code is a negative for many of our users. Adding more code size by default forces them to set a flag and be aware they need to set it; they really dislike us adding such flags over time.

This is fair. I do think we can get a lot from a few bytes of JSON in the JS payload (<100?).

But maybe I can "purchase" some bytes by pointing out a possible size reduction :))

I opened up an emscripten js ([this one](https://hoten.cc/zc/play/zelda.js), it's using a few months old toolchain) and what jumped out instantly was many `invoke_vii*` of various lengths

<img width="1219" alt="image" src="https://github.com/GoogleChrome/lighthouse/assets/4071474/06209e2b-4d2a-4778-99ee-1e3e92d3cc8f">
<img width="1037" alt="image" src="https://github.com/GoogleChrome/lighthouse/assets/4071474/3aeb0072-6807-45b5-bb8a-ddd159f3e2c6">

Maybe this compresses well, I don't know. But perhaps some of this could be generated at runtime like:

```js
// from
"nc": invoke_viiiiiiiiiiiiiiiiiiiiiiiiiiiiii,
// to
"nc": makeInvoke(30, 0) // or whatever

function makeInvoke(i, j) {
const dyncall = globalThis[`dynCall_${somehow generate those letters representing call signature idk}`]
return () => { ... }
}
```

Just considering the `invoke_*` methods, that's something like 16KB in this bundle.



> 2. Awkward incentives. We as toolchain makers would love to know how much our toolchain is used, but making our users ship larger builds for our benefit feels wrong. (In the context here, the user would benefit, so this is not entirely applicable.)



> 3. Lack of precedent in the JS space. (Has this changed?)

No - but that's largely because extended information is in an external resource (`.js.map`). Maybe we can put emscripten's meta information into a json file next to the source but with a `.json` extension? Then developer tools could know to look there if a emscripten WASM is detected on the page.

**Reply by kripken** [reply]
> or the very specific "was this built with NODE support", can we just look for ENVIRONMENT_IS_NODE variable being present?

Unfortunately, no, as closure compiler will minify that name.

About those invokes, yeah, that is wasteful! This is actually due to either exceptions or longjmp, and both can be avoided in modern browsers using the new wasm exceptions spec, so optimally that codebase would be rebuilt with it. (Wasm exceptions are fairly recent, so not everyone has heard of them or is ok with depending on it.) That's one reason we haven't heavily optimized that JS, but you are definitely right that we could in principle.

> No - but that's largely because extended information is in an external resource (.js.map). Maybe we can put emscripten's meta information into a json file next to the source but with a .json extension? Then developer tools could know to look there if a emscripten WASM is detected on the page.

I think an option to emit that file could be ok. But we wouldn't want it by default for the usual reasons. Perhaps it could be something like "build a release build + this flag locally, and then run devtools to get suggestions" - something like that?

**Comment by tomayac** [comment]
> * Lack of precedent in the JS space. (Has this changed?)

Actually a lot of JS libraries leave traces that tools like [Wappalyzer use to detect](https://github.com/wappalyzer/wappalyzer/tree/master/src/technologies) the presence of said libraries, including version numbers and sometimes more. For example, this is the tool's output for https://emscripten.org/:

![Screenshot 2023-08-08 at 18 51 10](https://github.com/GoogleChrome/lighthouse/assets/145676/cf7d4278-bdc9-45e1-8a97-825636790610)

**Reply by kripken** [reply]
Interesting! How does that work? Are those libraries not fully minified, or do they tell the minifier to leave those lines in intentionally? (like they do copyright lines, perhaps?)

**Reply by connorjclark** [reply]
They expose their info on the JS global object or their library objects directly. Here is another library that encapsulates getting version info from tons of libraries: https://github.com/johnmichel/Library-Detector-for-Chrome/blob/master/library/libraries.js#L1140C6-L1155

Every library I'm aware of does this always, with no configuration to disable.

**Reply by kripken** [reply]
I'm very curious about this. Is there a rationale explained somewhere? That is, what are the library maker's reasons for doing it?

I can see practical reasons for it - in that link, maybe Angular needs to communicate with plugins or such, so this is effectively an ABI of sorts. I'd be more interested if libraries do it in pure comments that have no practical effect on their own execution (which would be the case for us).

**Reply by connorjclark** [reply]
It's likely done universally in a web exposable way because it is a useful debugging tool when authors investigate reported issues with their libraries on live webpages -- and because the cost to add it is nearly trivial.

**Reply by kripken** [reply]
Interesting, thanks. Might be different for Emscripten, then, as we haven't needed it for that type of bug report situation.

Overall the difference might be that Emscripten is a compiler that happens to ship some library code (but mostly ships compiled user code) as opposed to a library itself. In the past when we looked for precedent in the Web space we compared Emscripten to other compilers and minifiers (like closure compiler, Uglify, Terser, etc.).

Overall I'm open-minded here, but making a change in Emscripten to increase code size for this reason would not be a simple process. If we think it's really worth trying I'm not opposed to that discussion, though.

**Reply by tomayac** [reply]
Given that [Wappalyzer already detects Wasm](https://github.com/wappalyzer/wappalyzer/blob/58d1dac114a5fcedfde8e1a23446f09b636f34c5/src/technologies/w.json#L772-L788), I think it’s not out of this world to mark the Emscripten glue code with meta information like the used Emscripten version, since there’s clearly interest outside of the Lighthouse issue we’re discussing here. There may be a clever way to encode important other information in a bit vector or so.

**Reply by kripken** [reply]
Sorry, I'm not familiar with Wappalyzer and I'm not sure what I should infer from that link (I just see a little JSON that for some reason mentions `cats: 27` :cat: :smile: ).

I think if you can find precedent in the minifier/compiler space that would be much more compelling for me personally. But, regardless, it's not going to be up to myself specifically, we'd need to see what the community thinks. We can either continue to discuss here, or we could open an Emscripten issue for that.

**Reply by tomayac** [reply]
FWIW, the fields [are documented](https://github.com/wappalyzer/wappalyzer/tree/58d1dac114a5fcedfde8e1a23446f09b636f34c5#json-fields). The idea of Wappalyzer is essentially what we're discussing here: analyzing the source code or network traces for telling patterns that let the tool find out if a certain technology is used (here: the header is `"Content-Type": "application/wasm"` and it has `wasmBinary` or `wasmBinaryFile` in its JavaScript). So I think before we start changing the way Emscripten works, we should continue exploring which of the "telling traces" approaches could work and look to remain stable in the future.

**Reply by kripken** [reply]
I see, thanks. Checking the content-type looks like the only robust thing there, as `wasmBinary[File]` will not exist in a fully minified JS file, and `wasm_exec.js` is a very specific thing to the main Go wasm toolchain. But the content-type should work for the great majority of cases (except for pages that JIT wasm at runtime, or download and decompress it in JS for some reason).
